### DISCUSSION TITLE
Handling exceptions from detached threads

### DISCUSSION THREAD

**Discussion by ptsouchlos** [ptsouchlos/thread-pool#44] [General]
From: @JuniorSuperTux
> Any thoughts on implementing API for getting exceptions from the detached threads?

**Comment by ptsouchlos** [comment]
First, I assume that by a detached thread, you mean like when you call `thread.detach()` right? 

If so, my question would be why you're interested in the exception state of a detached thread. Could you explain a bit about your use case?

**Comment by JuniorSuperTux** [comment]
I'm not sure if I learnt correctly, but there's ``boost::io_context`` in boost which allows callbacks to be run in an asynchronous manner, and what I guess I'll do is iterate through all the io_context and pass all the objects and io_context::run function pointer directly to the thread pool without ever calling get. And from what I've read io_context::run throws exceptions, and I would like to check them.

(I'm totally unsure whether I'm talking some nonsense bullshit rn)

**Reply by ptsouchlos** [reply]
I see. So, in the context of my thread pool library, you would do this through the `std::future` interface. When the thread pool executes the work, it captures exceptions (if any) and passes them to the `std::promise` and thus back to the caller through the returned `std::future` object. 

So you could do:

```cpp
// assume the task throws an exception
auto throw_future = pool.enqueue(throw_task, 1);
// the exception is caught and put into the future. 
// getting the future will re-throw the exception
auto result = throw_future.get();
```

Hopefully that makes sense and is what you're looking for.
